### PR TITLE
PERF: ArrowExtensionArray.__iter__

### PR DESCRIPTION
**Pull request by lukemanley**
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/v2.0.0.rst` file if fixing a bug or adding a new feature.

Faster iteration through pyarrow-backed arrays.

```
from asv_bench.benchmarks.strings import Iter

dtype = "string[pyarrow]"
bench = Iter()
bench.setup(dtype)

%timeit bench.time_iter(dtype)

# 266 ms ± 6.66 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)    <- main
# 57.7 ms ± 679 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)  <- PR
```

### CODE REVIEW

File: pandas/core/arrays/arrow/array.py [hunk 1]
```diff
@@ -330,6 +331,18 @@ def __getitem__(self, item: PositionalIndexer):
             else:
                 return scalar
 
+    def __iter__(self) -> Iterator[Any]:
+        """
+        Iterate over elements of the array.
+        """
+        na_value = self._dtype.na_value
+        for value in self._data:
```
Comment (mroeschke) [inline pandas/core/arrays/arrow/array.py:339]: Is there any benefit to combining the chunks of `self._data` and then iterating?

Comment (lukemanley) [reply pandas/core/arrays/arrow/array.py:339]: Maybe seeing a 5-10% improvement testing locally, but not much. Seems to be 5-10% with 10 chunks or 10000 chunks. I can add it if you want?

Comment (lukemanley) [reply pandas/core/arrays/arrow/array.py:339]: I don't think I'm seeing a consistent improvement at low number of chunks. 

If curious, here is the snippet I'm testing with:

```
import pyarrow as pa

def time_iter(arr):
    for v in arr:
        pass

num_chunks = 2
chunk_size = 10**5
data = [np.arange(chunk_size)] * num_chunks
arr = pa.chunked_array(data)

%timeit time_iter(arr)
%timeit time_iter(arr.combine_chunks())
```

Comment (jbrockmendel) [reply pandas/core/arrays/arrow/array.py:343]: IIUC doing combine_chunks will cause a copy, so any speed improvements will be offset by memory increase.  if so, i think not worth it

**Review by jbrockmendel — APPROVED** [review]

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @lukemanley!

**Pull request merged**